Fix Invalid foreign key when applying schema snapshot

PeeraJ opened the pull request:
Fix #9723 
it's my first time contributing back to Directus community please let me know if something needs to be changed. 
<img width="1058" alt="image" src="https://user-images.githubusercontent.com/12496997/158652367-bd190e83-281d-4daf-8e3f-9d8eef6e0abb.png">

as of before if you have data models as shown when you try to apply schema it will throw an error like this
<img width="1016" alt="image" src="https://user-images.githubusercontent.com/12496997/158653005-989dc79e-5187-4898-86a6-0f1c25ccfd26.png">

after fix
<img width="535" alt="image" src="https://user-images.githubusercontent.com/12496997/158653221-4aa84613-903e-4383-af50-3a027f67c967.png">

PeeraJ committed:
add logic to create parent collection before apply

PeeraJ committed:
recursively create collection

PeeraJ committed:
Recursively delete collection

PeeraJ committed:
cleanup

eyecatchup commented:
This is a real deal breaker. @rijkvanzanten Could you gimme a rough estimate when you will have time to review this PR? Thanks.

azrikahar commented on api/src/utils/apply-snapshot.ts:
```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
-					throw err;
+		const createCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
+				for (const item of collections) {
```
```suggestion
				for (const { collection, diff } of collections) {
```

More of a nitpick but let's keep the original destructuring to keep things similar as before 👍 (unless of course there's a naming collision that I'm missing here)

azrikahar commented on api/src/utils/apply-snapshot.ts:
```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
+		};
+		const deleteCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
+				for (const item of collections) {
```
```suggestion
				for (const { collection, diff } of collections) {
```

PeeraJ committed:
fix import

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

PeeraJ committed:
fix deconstruction of collections

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

PeeraJ committed:
fix deconstruction of collections

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

eyecatchup committed:
Fix destructing

eyecatchup commented:
I've added a [PR](https://github.com/PeeraJ/directus/pull/1/commits) to the patch branch to fix the deconstructing.

PeeraJ committed:
Merge pull request #1 from eyecatchup/patch-5

Fix destructing

eyecatchup commented:
I see there're some unanswered questions from your initial review. @azrikahar could you please clarify if you're waiting for further specific answers from @PeeraJ or if you just not had the time to review the updates to this PR?

azrikahar commented:
> could you please clarify if you're waiting for further specific answers from @PeeraJ or if you just not had the time to review the updates to this PR?

Honestly both! If PeeraJ is still busy and time permits on my end, I'll try to add the remaining tweaks and merge conflict before the next planned release. Though ultimately it can be rather hard to give any certainty with other things taking priority at the moment. Hope that's understandable. 👍

eyecatchup committed:
Move type declaration to the top

As requested during code review https://github.com/directus/directus/pull/12198#discussion_r857330627

eyecatchup commented:
Sure thing, understandable. Thanks for the quick reply! I've [moved the type declaration to the top](https://github.com/PeeraJ/directus/pull/2), hopefully @PeeraJ can merge this soon. Are there any remaining "tweaks" you would need? Otherwise I could take care of resolving the merge conflicts after my PR was merged to get this fixed asap, without blocking your time.

PeeraJ commented:
> Sure thing, understandable. Thanks for the quick reply! I've [moved the type declaration to the top](https://github.com/PeeraJ/directus/pull/2), hopefully @PeeraJ can merge this soon. Are there any remaining "tweaks" you would need? Otherwise I could take care of resolving the merge conflicts after my PR was merged to get this fixed asap, without blocking your time.

I'm so sorry to leave this unattended, I just came back from a long vacation and am unable to fix those problems right away. Thanks a lot for the pr though.🙏

PeeraJ committed:
Merge pull request #2 from eyecatchup/patch-5

Move type declaration to the top

PeeraJ committed:
remove unnecessary empty array check.

PeeraJ commented on api/src/utils/apply-snapshot.ts:
```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
-					logger.error(`Failed to delete collection "${collection}"`);
-					throw err;
+		const createCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
```
removed.

eyecatchup commented:
Great. Thanks @PeeraJ for merging my change and removing the unnecessary empty check. That being done, it only needs to resolve the merge conflicts. Since I can not do it on this PR, I needed to create a duplicate. This PR https://github.com/directus/directus/pull/13096 includes all the changes from PeeraJ and me and resolves the merge conflict. @azrikahar anything else I can assist with to get this merged?

azrikahar committed:
Merge branch 'main' into fix/schema-apply-parent-collection

azrikahar commented:
@PeeraJ no worries! Thanks for updating promptly. I'll take it from here if that's okay with you 👍

> anything else I can assist with to get this merged?

@eyecatchup all good. I'd like to lean on to the safer side and do some tests on my end just to prevent any oversight here. I've resolved the merge conflict for now, and I'll add on any changes if needed before wrapping this up. Thanks again for helping out.

azrikahar committed:
Merge branch 'main' into fix/schema-apply-parent-collection

azrikahar committed:
remove redundant map functions

azrikahar committed:
replace lodash functions in favor of built-in ones

azrikahar committed:
tweaks

azrikahar committed:
organize imports
